UI/MapWnd.cpp

```diff
@@ -2778,57 +2781,147 @@ void MapWnd::ClearSystemRenderingBuffers() {
-                ancestor[newSys] = sysID;
-                if (newSys==end_sys_it) {
-                    int iSys = newSys;
-                    while ((ancestor[iSys] !=-1)&&( ancestor[iSys] != iSys )) {
-                        path.push_back(iSys);
-                        iSys = ancestor[iSys];
+namespace {
+    // SupplyLaneMap map keyed by system containing all system
+    // corresponding to valid supply lane destinations
+    typedef boost::unordered_multimap<int,int> SupplyLaneMMap;
+
+    // AncestorInfo stores the \p ids of systems one back on the path.
+    struct AncestorInfo {
+        AncestorInfo(int a, int o) : ids(1, a), single_origin(o) {}
+        std::vector<int> ids;  // previous systems
+        // originating terminal or nothing for connected resources
+        boost::optional<int> single_origin;
+    };
+
+    struct PrevCurrInfo {
+        PrevCurrInfo(int p, int n, int o) : prev(p), curr(n), origin(o) {}
+        int prev, curr, origin;
+    };
+
+    /**
+       Returns either:
+       a) a \p good_path between from any \p terminal_points to all
+       reachable \p terminal_points along the \p supply_lanes.
+       b) empty indicating no more work to do.
+
+     */
+    void GetPathsThroughSupplyLanes(
+        boost::unordered_set<int> & good_path,
+        const boost::unordered_set<int> & terminal_points,
+        const SupplyLaneMMap& supply_lanes)
+    {
+        good_path.clear();
+
+        // No terminal points, so all paths lead nowhere.
+        if (terminal_points.empty())
+            return;
+
+        boost::unordered_map<int, AncestorInfo> ancestors;
+        std::deque<PrevCurrInfo> tryNext;
+        std::vector<int> path;
+
+        // A list of all reachable midpoints between at least two different terminal points
+        std::vector<int> reachable_midpoints;
+
+        // Start from all terminal points.
+        for (boost::unordered_set<int>::const_iterator start_it = terminal_points.begin();
+             start_it != terminal_points.end(); ++start_it) {
+            tryNext.push_back(PrevCurrInfo(*start_it, *start_it, *start_it));
+            ancestors.insert(std::make_pair(*start_it, AncestorInfo(*start_it, *start_it)));
+        }
+
+        // Find all reachable midpoints where paths from two different
+        // terminal points meet.
+        while (!tryNext.empty() ) {
+            const PrevCurrInfo & curr = tryNext.front();
+
+            std::pair<SupplyLaneMMap::const_iterator, SupplyLaneMMap::const_iterator> supplylane_endpoints
+                = supply_lanes.equal_range(curr.curr);
+            for (SupplyLaneMMap::const_iterator sup_it = supplylane_endpoints.first;
+                 sup_it != supplylane_endpoints.second; ++sup_it)
+            {
+                int next = sup_it->second;
+                if (next == curr.prev)
+                    continue;
+                boost::unordered_map<int, AncestorInfo>::iterator ancestor = ancestors.find(next);
```

Comment:
Yes.

`unordered_set`, `unordered_map` and their `multimap` versions have average worst case linear complexity, `O(1)`, when inserting, searching, deleting, whereas `set`, `map` and their `multimap` version have logarithmic complexity <code>O(log<sub>2</sub>N)</code> for the same operations.  That is the price of `set` and `map` being ordered.

So assuming that the supply plus terminal point network has `N = 64` systems in it.  Every insert, find and delete will take on average <code>log<sub>2</sub>(64)</code> = 6 times longer if I use `set` instead of `unordered_set`.

`unordered_map` and `unordered_set` are usually the better option if you don't need the sorting behavior of `set` and `map`.